agent/pool/pool.go

```diff
@@ -372,62 +377,54 @@ func (p *ConnPool) dial(
 	return conn, hc, nil
 }
 
-// DialTimeoutWithRPCTypeViaMeshGateway dials the destination node and sets up
-// the connection to be the correct RPC type using ALPN. This currently is
-// exclusively used to dial other servers in foreign datacenters via mesh
-// gateways.
-//
-// NOTE: There is a close mirror of this method in agent/consul/wanfed/wanfed.go:dial
-func DialTimeoutWithRPCTypeViaMeshGateway(
-	dc string,
-	nodeName string,
-	addr net.Addr,
-	src *net.TCPAddr,
-	wrapper tlsutil.ALPNWrapper,
-	actualRPCType RPCType,
-	tlsRPCType RPCType,
-	// gateway stuff
+// DialRPCViaMeshGateway dials the destination node and sets up the connection
+// to be the correct RPC type using ALPN. This currently is exclusively used to
+// dial other servers in foreign datacenters via mesh gateways.
+func DialRPCViaMeshGateway(
+	ctx context.Context,
+	dc string, // (metadata.Server).Datacenter
+	nodeName string, // (metadata.Server).ShortName
+	srcAddr *net.TCPAddr,
+	alpnWrapper tlsutil.ALPNWrapper,
+	nextProto string,
 	dialingFromServer bool,
-	tlsConfigurator *tlsutil.Configurator,
 	gatewayResolver func(string) string,
-	thisDatacenter string,
 ) (net.Conn, HalfCloser, error) {
 	if !dialingFromServer {
 		return nil, nil, fmt.Errorf("must dial via mesh gateways from a server agent")
 	} else if gatewayResolver == nil {
 		return nil, nil, fmt.Errorf("gatewayResolver is nil")
-	} else if tlsConfigurator == nil {
-		return nil, nil, fmt.Errorf("tlsConfigurator is nil")
-	} else if dc == thisDatacenter {
-		return nil, nil, fmt.Errorf("cannot dial servers in the same datacenter via a mesh gateway")
-	} else if wrapper == nil {
+	} else if alpnWrapper == nil {
 		return nil, nil, fmt.Errorf("cannot dial via a mesh gateway when outgoing TLS is disabled")
 	}
 
-	nextProto := actualRPCType.ALPNString()
-	if nextProto == "" {
-		return nil, nil, fmt.Errorf("rpc type %d cannot be routed through a mesh gateway", actualRPCType)
-	}
-
 	gwAddr := gatewayResolver(dc)
 	if gwAddr == "" {
 		return nil, nil, structs.ErrDCNotAvailable
 	}
 
-	dialer := &net.Dialer{LocalAddr: src, Timeout: defaultDialTimeout}
+	var dialTimeout time.Duration
+	if nextProto != ALPN_RPCGRPC && dialTimeout == 0 {
```

Comment:
```suggestion
	if nextProto != ALPN_RPCGRPC {
```
Won't dialTimeout always be 0?